backend/remote-state/http/backend_state.go

```diff
@@ -0,0 +1,177 @@
+package http
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/hashicorp/terraform/backend"
+	"github.com/hashicorp/terraform/state"
+	"github.com/hashicorp/terraform/state/remote"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+const (
+	stateFileSuffix = ".tfstate"
+	lockFileSuffix  = ".tflock"
+)
+
+// States get a list of all states.
+func (b *Backend) States() ([]string, error) {
+	var result []string
+	client := &RemoteClient{
+		client:   b.client,
+		address:  b.address,
+		username: b.username,
+		password: b.password,
+	}
+
+	resp, err := client.Get()
+	if err != nil {
+		return nil, err
+	}
+	// Read in the body
+	buff := string(resp.Data)
+	// Get all data and make it a slice.
+	// The REST API should return all the files from the root of the project name.
+	// This means we get all the state files of a project
+	// ex. buff := "default.tfstate default.tflock foo.tfstate bar.tfstate"
+	buff = strings.Replace(buff, ",", " ", -1)
+	// make it a slice so we can search and loop thru it
+	buffSlice := strings.Fields(buff)
+	for _, el := range buffSlice {
+		// iterate on response and make sure we only get .tfstate files
+		// This can be implemented in REST API but just to be safe.
+		// get the base name
+		fName := filepath.Base(el)
+		// get the extension
+		extName := filepath.Ext(el)
+		// separate them
+		bname := fName[:len(fName)-len(extName)]
+		if extName == stateFileSuffix {
+			// append the state to the result
+			result = append(result, bname)
+		}
+	}
+	// check if we already have default named state and add it otherwise
+	for _, n := range result {
```

Comment:
If you don't want to loop back through the `result` (you looped through it once above on L44), you can just have a `bool` that tracks if we encountered the default state name in the above loop. Then after the loop is completed, you could replace this second loop with just an `if` and append.

Reply:
I wanted to use binary search but I didn't manage to make it work, so I went for another loop

Reply:
I am saying you can save a loop doing something like this:

```
// tracks whether or not we've seen the default state
foundDefaultState := false
for _, el := range buffSlice {
  // not a TF state file, skip
  extName := filepath.Ext(el)
  if extName != stateFileSuffix {
    continue
  }
  // basename of the file with extension
  fName := filepath.Base(el)
  // remove the extension to get file name, add to list of results
  file := fName[:len(fName) - len(extName)]
  result = append(rest, file)
  // we found the default state file
  if file == backend.DefaultStateName {
    foundDefaultState = true
  }
}
// did not encounter the default state, add it now
if !foundDefaultState {
  return append(result, backend.DefaultStateName)
}
return result
```

You save the second iteration of the loop since you track whether or not you've sen it when you're doing the first loop